semgrep-core/src/engine/Match_rules.ml

```diff
@@ -95,25 +95,25 @@ let check ~match_hook ~timeout ~timeout_threshold default_config rules xtarget =
              else true
            in
            if not relevant_rule then (
-             logger#trace "skipping rule %s for %s" (fst r.R.id) file;
+             logger#trace "skipping rule %s for %s" (R.id r |> fst) file;
              Right r)
            else
-             let rule_id = fst r.R.id in
+             let rule_id = R.id r |> fst in
              Rule.last_matched_rule := Some rule_id;
              Common.profile_code (spf "real_rule:%s" rule_id) (fun () ->
                  let match_result =
                    timeout_function r file timeout (fun () ->
                        (* dispatching *)
-                       match r.R.mode with
-                       | Search pformula ->
+                       match r with
+                       | `Search _ as r ->
```

Comment:
You don't need `as r` here because `r` is already bound to the rule

Reply:
You do, because otherwise the compiler doesn't know we're in the specific variant.